### PR TITLE
PLANNER-2114 Extract build methods from Config

### PR DESCRIPTION
**Pull request by rsynek**
The scope of this PR is move selectors.

**JIRA**: [PLANNER-2114](https://issues.redhat.com/browse/PLANNER-2114)

**referenced Pull Requests**: None

**How to retest or run**:

* a pull request please add comment: regex **[.\*[j|J]enkins,?.\*(retest|test) this.\*]**

* a full downstream build please add comment: regex **[.*\[j|J]enkins,?.\*(execute|run|trigger|start|do) fdb.\*]**

* a compile downstream build please  add comment: regex **[.\*[j|J]enkins,?.\*(execute|run|trigger|start|do) cdb.\*]**

* a full production downstream please add comment: regex **[.\*[j|J]enkins,?.\*(execute|run|trigger|start|do) product fdb.\*]**

* an upstream build please add comment: regex **[.\*[j|J]enkins,?.\*(execute|run|trigger|start|do) upstream.\*]**

i.e for running a full downstream build =  **Jenkins do fdb**

### CODE REVIEW

**Comment by rsynek** [comment]
I reviewed the test coverage and don't see how to improve it further without testing fail-fast checks (already did for some more complex checks).

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java [hunk 1]
```diff
@@ -727,6 +727,32 @@ public ScoreDefinition getScoreDefinition() {
     // Model methods
     // ************************************************************************
 
+    public EntityDescriptor deduceEntityDescriptor(Class<?> entityClass) {
```
[the remark this answers is not in the document]
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:730]: Please don't access collections directly. Leaking references to collections leads to a virtual impossibility of figuring out who has the opportunity to read (and more importantly write) it.

The config code is full of places where collections are being exported to other places, where they are being modified indiscriminately. Refactoring such code becomes a nightmare, as you are never sure if you've covered all the possible escape routes. 

Therefore I would like to suggest that exporting a modifiable collection should be prohibited.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactory.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     * Builds {@link MoveSelector} from the {@link MoveSelectorConfig} and provided parameters.
+     * 
+     * @param configPolicy never null
+     * @param minimumCacheType never null, If caching is used (different from {@link SelectionCacheType#JUST_IN_TIME}),
+     *        then it should be at least this {@link SelectionCacheType} because an ancestor already uses such caching
+     *        and less would be pointless.
+     * @param inheritedSelectionOrder never null
+     * @return never null
+     */
+    MoveSelector buildMoveSelector(HeuristicConfigPolicy configPolicy, SelectionCacheType minimumCacheType,
```
[the remark this answers is not in the document]
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactory.java:91]: This is an interface, maybe you meant the same method on the `AbstractMoveSelectorFactory`?. Added, thanks.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:730]: As for the first suggestion, I will look into it in the next PR, which will be focused on the entity and value selection config. As for the second suggestion, I would end up duplicating a part of the `getGenuineEntityDescriptors` method, with the drawbacks @triceo has mentioned (although the reference does not leak out the class).

@yurloc could you please explain what is your concern about not accessing the map directly?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java [hunk 3]
```diff
@@ -516,6 +516,58 @@ public boolean hasAnyChainedGenuineVariables() {
...
+                        + ") has no configured variableName (" + variableName
+                        + ") for entityClass (" + getEntityClass()
+                        + ") and because there are multiple variableNames ("
+                        + getGenuineVariableNameSet()
+                        + "), it cannot be deduced automatically.");
+            }
+            variableDescriptor = variableDescriptors.iterator().next();
+        }
+        return variableDescriptor;
+    }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java:543]: This method, it's actually two methods. One accepts the argument, the other doesn't need it. Let's do it like that.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java [hunk 4]
```diff
@@ -516,6 +516,58 @@ public boolean hasAnyChainedGenuineVariables() {
...
+
+    public List<GenuineVariableDescriptor<Solution_>> deduceVariableDescriptorList(List<String> variableNameIncludeList) {
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java:545]: I'd prefer a no-arg overload to just putting `null` here.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java [hunk 5]
```diff
@@ -727,6 +727,32 @@ public ScoreDefinition getScoreDefinition() {
...
+                        + ") has no entityClass (" + entityClass
+                        + ") configured and because there are multiple in the entityClassSet ("
+                        + getEntityClassSet()
+                        + "), it cannot be deduced automatically.");
+            }
+            entityDescriptor = entityDescriptors.iterator().next();
+        }
+        return entityDescriptor;
+    }
+
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:755]: Same comment on the `entityClass` nullity. This is two methods masquerading as one.

File: optaplanner-core/src/test/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    @Test
+    void stepRandom() {
+        final MoveSelector baseMoveSelector = SelectorTestUtils.mockMoveSelector(DummyMove.class);
+        DummyMoveSelectorConfig moveSelectorConfig = new DummyMoveSelectorConfig();
+        MoveSelectorFactory moveSelectorFactory =
+                new AssertingMoveSelectorFactory(moveSelectorConfig, baseMoveSelector, SelectionCacheType.STEP, false);
+        moveSelectorConfig.setCacheType(SelectionCacheType.STEP);
+        moveSelectorConfig.setSelectionOrder(SelectionOrder.RANDOM);
+        MoveSelector moveSelector = moveSelectorFactory.buildMoveSelector(
+                buildHeuristicConfigPolicy(),
```
Comment (triceo) [inline optaplanner-core/src/test/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactoryTest.java:121]: Is this what the code formatter does? Pretty sure it fits on one 120-char line.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java (hunk 5 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:755]: Regarding the last three comments - I agree, my plan was to look into it in the next PR, which focuses on the remaining part of the config code - including config of `ValueSelector` and `EntitySelector`. For this PR, I just moved these existing methods to - in my opinion - a more suitable place without changing their internals.

I should have clearly stated my intentions before opening the PR for review.

File: optaplanner-core/src/test/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactoryTest.java (hunk 6 above)
Comment (rsynek) [reply optaplanner-core/src/test/java/org/optaplanner/core/impl/heuristic/selector/move/MoveSelectorFactoryTest.java:121]: No, the formatter is not guilty; there must have been a new line already and the formatter didn't touch it, precisely as we wanted. Some of the test methods were moved from a different test class.

Good catch, thanks!

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java (hunk 3 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java:543]: will be addressed in the next PR, see the comment below.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java (hunk 4 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/entity/descriptor/EntityDescriptor.java:545]: will be addressed in the next PR, see the comment below.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yurloc) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:730]: The original post in this thread was a note for myself. I understand the second suggestion might be a bit confusing. Let's continue on line 733.

[the remark this answers is not in the document]
Comment (yurloc) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:730]: I fully agree with what Lukas have said.

**Pull request merged**